Fix discussion user roles

awaisdar001 opened the pull request:
## [TNL-5816](https://openedx.atlassian.net/browse/TNL-5816)

### Description

When a student visits any users' discussion profile, his roles are displayed on the page. I have done the following changes in this PR.

- I have moved the logic of getting users' roles in the views.
- I have fixed the logic of getting users' roles is now with respect to the course 
- I am now raising `Http404` if the user is not enrolled in the course.

https://github.com/edx/edx-platform/pull/13939/files



### How to Test?

**Stage**: 
- In the discussion forum, visit a learner profile page. [Aishaq discussion profile](https://courses.stage.edx.org/courses/course-v1:HarvardX+CS101+2016_T4/discussion/forum/users/5183691). Note down the learner roles listed under the username as `Community TA, Moderator, Student`
- Verify the learner roles from the [Instructor Dashboard - Course-A](https://courses.stage.edx.org/courses/course-v1:HarvardX+CS101+2016_T4/instructor#view-membership). Verify the learner has only `Moderator` role in the course. 

- Go to the [Instructor Dashboard - Course-B](https://courses.stage.edx.org/courses/course-v1:HarvardX+CS102+2016_T4/instructor#view-membership) and verify the learner has `Community TA` role 


**Sandbox**
- [https://awais.sandbox.edx.org/](https://awais.sandbox.edx.org/)
- Create two courses (Course-A) and (Course-B) in the studio.
- Enroll a learner (audit) in both courses.
- Goto instructor dashboard of Course-A and add a role `Moderator` for the learner.
- Goto instructor dashboard of Course-B and add a role `Community TA` for the learner.
- Go to the discuss tab of Course-A and open learner profile and verify the learner has only `Moderator` role listed. 
- Go to the discuss tab of Course-B and open learner profile and verify the learner has only `Community TA` role listed. E.g.(`www.courses.edx.org/course-id-here/discussion/forum/users/learner-id/)

### Testing
- [x] Acceptance tests
- [x] Unit tests


### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] F.I.R: @noraiz-anwar
- [x] Code review: @attiyaIshaque 
- [x] Code review: @Qubad786 

FYI: @adampalay 

### Post-review
- [x] Rebase and squash commits

awaisdar001 commented:
jenkins run jenkins

noraiz-anwar commented on common/test/acceptance/tests/discussion/test_discussion.py:
```diff
@@ -1145,21 +1145,29 @@ class DiscussionUserProfileTest(UniqueCourseTest):
-        ).visit().get_user_id()
+        self.user_id = self.setup_user()
+
+    def setup_course(self):
```
I wonder if we really need this method here. what if we can use one liner return at the place where we are calling this function `setup_course`.

awaisdar001 replied:
If you still think we can, please let me know why? I am open to listen your suggestion.

noraiz-anwar replied:
@awaisdar001 I just saw you used your `setup_course` at more than one place, like one at line number 1296. It is fair enough to avoid duplication by making a method.
Satisfied.

awaisdar001 commented on common/test/acceptance/tests/discussion/test_discussion.py:
```diff
@@ -1262,6 +1270,43 @@ def test_redirects_to_learner_profile(self):
+        page = self.check_pages(1)
+        student_roles = page.get_user_roles()
+        self.assertTrue(
+            all([role in student_roles for role in expected_student_roles])
```
You can say, I am verifying the roles on the page are equal to the expected roles. 
i.e. page has listed all the rols. `['Administrator', 'Community TA', 'Moderator', 'Student']`

awaisdar001 replied:
btw, i have changed the `assert` to be much simpler now. thank you for brining it up.

attiyaIshaque commented on lms/templates/discussion/_user_profile.html:
```diff
@@ -3,8 +3,7 @@
 <div class="user-profile">
   <div class="sidebar-username"><a class="learner-profile-link" href="${learner_profile_page_url}">${django_user.username | h}</a></div>
   <div class="sidebar-user-roles">
-    <% role_names = django_user.roles.order_by("name").values_list("name", flat=True).distinct() %>
```
I am confused about this line. why did you remove?

awaisdar001 replied:
Perfectly fine. I can help you understand. 

Actually, I moved the roles logic in the view and passed from there. Check `user_roles` context variable.

Qubad786 commented:
@awaisdar001 , Fix looks good, I would like to test it, if you are building sandbox let me know, otherwise can test locally.

awaisdar001 commented:
@Qubad786 I am building a sandbox, I will update once that's' done. I have updated the ticket though and added details how you can test on stage (current behavior) and sandbox (for the fix).

noraiz-anwar commented:
@awaisdar001 I am unable to open 'Link' provided under sandbox in PR description. It is returning http404.

awaisdar001 commented:
@noraiz-anwar Let me take a look.

noraiz-anwar commented:
@awaisdar001 yeah sure. Code looks perfect. Just need to test it on sandbox. 👍 after that.

noraiz-anwar commented:
jenkins run bokchoy

attiyaIshaque commented:
After testing on sandbox; it looks good to me. 
Once commits squashed 👍

awaisdar001 commented:
@attiyaIshaque Thanks for the quick review. appreciate that.

noraiz-anwar commented:
👍

Qubad786 commented:
Nice tests 👍 !

awaisdar001 commented:
I have rebased and squash the commits, going to merge this once the tests pass. Great work team for `M.I.C`
e.g. Making It Count.

adampalay commented:
look at that turnaround time :) :) :)
